### PR TITLE
Multihash CID implementation

### PR DESCRIPTION
**Pull request by angrybayblade**
## Proposed changes

This PR implements multihash CID to support v1 encodings for IPFS hashes

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes and CI passes too
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: aea/helpers/cid.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+        """Dunder to check object equivalence."""
+
+        if not isinstance(other, BaseCID):
+            return NotImplemented
+
+        return (
+            (self.version == other.version)
+            and (self.codec == other.codec)
+            and (self.multihash == other.multihash)
+        )
```
Comment (Karrenbelt) [inline aea/helpers/cid.py:103]: `return self.__dict__ == other.__dict__`

File: aea/helpers/cid.py [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+
+
+def to_v0(hash_string: str) -> str:
+    """Convert CID v1 hash to CID v0"""
+
+    cid = CID.from_string(hash_string)
+    if cid.version == 0:
+        raise ValueError(f"{hash_string} is already v0.")
+
+    return cast(CIDv1, cid).to_v0().encode().decode()
```
Comment (Karrenbelt) [inline aea/helpers/cid.py:267]: add encoding parameter as in `to_v1`, it is also used in `CIDv0.encode`

File: tests/test_helpers/test_cid.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+    with pytest.raises(
+        ValueError,
+        match="QmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR is already v0.",
+    ):
+        to_v0(HASH_V0)
+
+    with pytest.raises(
+        ValueError,
+        match="bafybeigdyrzt5sfp7udm7hu76uh7y26nf3efuylqabf3oclgtqy55fbzdi is already v1.",
```
Comment (Karrenbelt) [inline tests/test_helpers/test_cid.py:45]: here and right above `match=f"{HASH_V1} is already v1."`

**Review by Karrenbelt — COMMENTED** [review]
Looks good but - as in the other PRs - need to get CI to run and pass when we resolve / skip the pip lock

**Review by 8ball030 — APPROVED** [review]

File: aea/helpers/cid.py (hunk 2 above)
Comment (angrybayblade) [reply aea/helpers/cid.py:267]: v0 used `base58` encoding by default and I think it only supports one encoding format so we don't need encoding parameter here.

**Review by DavidMinarsch — APPROVED** [review]

**Pull request merged**